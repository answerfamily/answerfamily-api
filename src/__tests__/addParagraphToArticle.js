jest.mock('../dataloaders/auth0UserLoaderFactory');

const {
  loadESFixtures,
  unloadESFixtures,
  loadMongoFixtures,
  unloadMongoFixtures,
} = require('../../test/fixtures');
const gql = require('../../test/gql');
const {
  ES_FIXTURES,
  MONGO_FIXTURES,
} = require('../__fixtures__/articleParagraphReply');
const esClient = require('../lib/esClient');
const mongoClient = require('../lib/mongoClient');

describe('addParagraphToArticle', () => {
  beforeAll(async () => {
    await loadESFixtures(ES_FIXTURES);
    await loadMongoFixtures(MONGO_FIXTURES);
  });

  afterAll(async () => {
    await unloadESFixtures(ES_FIXTURES);
    await unloadMongoFixtures(MONGO_FIXTURES);
  });

  it('creates paragraph and paragraphReplies', async () => {
    const text = '明年五月台灣將成為亞洲第一個同性可合法結婚的國家';

    const { errors, data } = await gql`
      mutation($articleId: String!, $paragraph: ParagraphInput!) {
        addParagraphToArticle(articleId: $articleId, paragraph: $paragraph) {
          paragraphs {
            id
            text
            user {
              id
            }
            paragraphReplies {
              reply {
                id
              }
            }
          }
        }
      }
    `(
      {
        articleId: 'article1',
        paragraph: { text, replyIds: ['r1'] },
      },
      {
        userPromise: Promise.resolve({ iss: 'user-id' }),
      }
    );

    expect(errors).toBeUndefined();
    expect(
      // ID is random, thus exclude from snapshot
      // eslint-disable-next-line no-unused-vars
      data.addParagraphToArticle.paragraphs.map(({ id, ...data }) => data)
    ).toMatchSnapshot();

    const newParagraph = data.addParagraphToArticle.paragraphs.find(
      p => p.text === text
    );

    // cleanup
    const { db } = await mongoClient;
    await esClient.delete({
      index: 'paragraphs',
      type: '_doc',
      id: newParagraph.id,
    });
    await db
      .collection('paragraphReplies')
      .deleteMany({ paragraphId: newParagraph.id });
  });

  it('blocks non-logged in users', async () => {
    const { errors } = await gql`
      mutation($articleId: String!, $paragraph: ParagraphInput!) {
        addParagraphToArticle(articleId: $articleId, paragraph: $paragraph) {
          paragraphs {
            id
          }
        }
      }
    `({
      articleId: 'article1',
      paragraph: { text: 'haha', replyIds: ['r1'] },
    });

    // Should be unauthorzied
    expect(errors).toMatchSnapshot();
  });
});
