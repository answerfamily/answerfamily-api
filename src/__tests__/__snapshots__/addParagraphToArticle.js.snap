// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addParagraphToArticle blocks non-logged in users 1`] = `
Array [
  [GraphQLError: Must log in first],
]
`;

exports[`addParagraphToArticle connects reply and paragraph 1`] = `
Array [
  Object {
    "paragraphReplies": Array [
      Object {
        "reply": Object {
          "id": "r1",
        },
      },
    ],
    "text": "外國同性戀藉結婚來台治療愛滋病，都是健保支付，全民買單",
    "user": Object {
      "id": "author1",
    },
  },
  Object {
    "paragraphReplies": Array [
      Object {
        "reply": Object {
          "id": "r2",
        },
      },
    ],
    "text": "學校對國中小實施同志教育，將成為合法",
    "user": Object {
      "id": "author1",
    },
  },
  Object {
    "paragraphReplies": Array [
      Object {
        "reply": Object {
          "id": "r1",
        },
      },
    ],
    "text": "明年五月台灣將成為亞洲第一個同性可合法結婚的國家",
    "user": Object {
      "id": "user-id",
    },
  },
]
`;
